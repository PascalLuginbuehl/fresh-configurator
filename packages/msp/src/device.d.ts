export interface VoltageMeters {
  id: number;
  voltage: number;
}

export interface AnalogValues {
  voltage: number;
  mahDrawn: number;
  rssi: number;
  amperage: number;
}

export interface RawGpsData {
  fix: number;
  numSat: number;
  lat: number;
  lon: number;
  alt: number;
  speed: number;
  groundCourse: number;
}

export interface BoardInfo {
  boardIdentifier: string;
  boardVersion: number;
  boardType: number;
  targetCapabilities: number;
  targetName: string;
  boardName: string;
  manufacturerId: string;
  signature: number[];
  mcuTypeId: number;
  configurationState: number | undefined;
  sampleRateHz: number | undefined;
}

export type ImuUnit = [number, number, number];

export interface ImuData {
  accelerometer: ImuUnit;
  gyroscope: ImuUnit;
  magnetometer: ImuUnit;
}

export interface Kinematics {
  roll: number;
  pitch: number;
  heading: number;
}

export interface Status {
  cycleTime: number;
  i2cError: number;
  activeSensors: number;
  mode: number;
  profile: number;
}

export interface ExtendedStatus extends Status {
  cpuload: number;
  numProfiles: number;
  rateProfile: number;
  armingDisableCount?: number;
  armingDisableFlags?: number;
}

export interface RcTuning {
  rcRate: number;
  rcExpo: number;
  rollPitchRate: number;
  pitchRate: number;
  rollRate: number;
  yawRate: number;
  dynamicThrottlePid: number;
  throttleMid: number;
  throttleExpo: number;
  dynamicThrottleBreakpoint: number;
  rcYawExpo: number;
  rcYawRate: number;
  rcPitchRate: number;
  rcPitchExpo: number;
  throttleLimitType: number;
  throttleLimitPercent: number;
  rollRateLimit: number;
  pitchRateLimit: number;
  yawRateLimit: number;
}

export interface RcDeadband {
  deadband: number;
  yawDeadband: number;
  altHoldDeadhand: number;
  deadband3dThrottle: number;
}

export enum OSD_VIDEO_TYPES {
  AUTO,
  PAL,
  NTSC
}

export const OSD_VIDEO_VALUE_TO_TYPE = [
  OSD_VIDEO_TYPES.AUTO,
  OSD_VIDEO_TYPES.PAL,
  OSD_VIDEO_TYPES.NTSC
];

export enum OSD_UNIT_TYPES {
  IMPERIAL,
  METRIC
}

export const OSD_UNIT_VALUE_TO_TYPE = [
  OSD_UNIT_TYPES.IMPERIAL,
  OSD_UNIT_TYPES.METRIC
];

export enum OSD_PRECISION_TYPES {
  SECOND,
  HUNDREDTH,
  TENTH
}
export const OSD_PRECISION_VALUE_TO_TYPE = [
  OSD_PRECISION_TYPES.SECOND,
  OSD_PRECISION_TYPES.HUNDREDTH,
  OSD_PRECISION_TYPES.TENTH
];
