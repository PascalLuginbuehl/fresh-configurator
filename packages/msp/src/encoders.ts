import { crc8_dvb_s2_data } from "./utils";

export const encode_message_v1 = (
  code: number,
  data?: number[]
): ArrayBuffer => {
  let bufferOut;
  // always reserve 6 bytes for protocol overhead !
  if (data) {
    var size = data.length + 6,
      checksum = 0;

    bufferOut = new ArrayBuffer(size);
    let bufView = new Uint8Array(bufferOut);

    bufView[0] = 36; // $
    bufView[1] = 77; // M
    bufView[2] = 60; // <
    bufView[3] = data.length;
    bufView[4] = code;

    checksum = bufView[3] ^ bufView[4];

    for (var i = 0; i < data.length; i++) {
      bufView[i + 5] = data[i];

      checksum ^= bufView[i + 5];
    }

    bufView[5 + data.length] = checksum;
  } else {
    bufferOut = new ArrayBuffer(6);
    let bufView = new Uint8Array(bufferOut);

    bufView[0] = 36; // $
    bufView[1] = 77; // M
    bufView[2] = 60; // <
    bufView[3] = 0; // data length
    bufView[4] = code; // code
    bufView[5] = bufView[3] ^ bufView[4]; // checksum
  }
  return bufferOut;
};

export const encode_message_v2 = (
  code: number,
  data?: number[]
): ArrayBuffer => {
  const dataLength = data ? data.length : 0;
  // 9 bytes for protocol overhead
  const bufferSize = dataLength + 9;
  const bufferOut = new ArrayBuffer(bufferSize);
  const bufView = new Uint8Array(bufferOut);
  bufView[0] = 36; // $
  bufView[1] = 88; // X
  bufView[2] = 60; // <
  bufView[3] = 0; // flag
  bufView[4] = code & 0xff;
  bufView[5] = (code >> 8) & 0xff;
  bufView[6] = dataLength & 0xff;
  bufView[7] = (dataLength >> 8) & 0xff;
  for (let ii = 0; ii < dataLength; ii++) {
    bufView[8 + ii] = data![ii];
  }
  bufView[bufferSize - 1] = crc8_dvb_s2_data(bufView, 3, bufferSize - 1);
  return bufferOut;
};
