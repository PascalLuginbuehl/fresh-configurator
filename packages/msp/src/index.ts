export * from "./device";
export {
  ports,
  connections,
  isOpen,
  open,
  close,
  bytesRead,
  bytesWritten,
  packetErrors,
  apiVersion
} from "./serial/connection";
