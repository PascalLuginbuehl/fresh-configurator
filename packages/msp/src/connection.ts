/**
 * MSP Serial connections. This allows us to make multiple
 * serial connections, and enables execution of commands
 * on the given connection.
 *
 * The module handles making requests, and dealing with only
 * making one of the same request at a time.
 */

import SerialPort from "@serialport/stream";
import debug from "debug";
import MspDataView from "./dataview";
import { MspMessage, MspParser } from "./parser";
import { encodeMessageV2, encodeMessageV1 } from "./encoders";

import {
  Connection,
  MspCommand,
  ConnectionOptions,
  OpenConnectionFunction,
  OnCloseCallback,
} from "./types";

// Import bindings when we actually need them, so that libraries
// can import this library without the bindings needing to be available
// at import time
const initialiseBindings = async (): Promise<void> => {
  const { default: Binding } = await import("@serialport/bindings");
  SerialPort.Binding = Binding;
};

const log = debug("connection");

const connectionsMap: Record<string, Connection | undefined> = {};

/**
 * Execute the given MspCommand on the given port.
 *
 * Return the data received from the command
 */
export const execute = async (
  port: string,
  { code, data, timeout = 2500 }: MspCommand
): Promise<MspDataView> => {
  const connection = connectionsMap[port];
  if (!connection) {
    throw new Error(`${port} is not open`);
  }

  const sendData = data ? Buffer.from([...data]) : undefined;

  const { parser, serial, requests } = connection;

  const request =
    code > 254
      ? encodeMessageV2(code, sendData)
      : encodeMessageV1(code, sendData);

  const requestKey = request.toString("utf-8");
  let dataRequest = requests[requestKey];

  // Prevent FC lockup by checking if this request is currently being performed
  // and if not, making the request
  if (!dataRequest) {
    dataRequest = new Promise((resolve, reject) => {
      // Throw an error if timeout is reached
      const timeoutId = setTimeout(() => {
        delete requests[requestKey];
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        parser.off("data", onData);
        reject(new Error("Timed out during execution"));
      }, timeout);

      function onData(message: MspMessage): void {
        if (message.code === code) {
          if (message.crcError && connection) {
            connection.packetErrors += 1;
            return;
          }
          log(
            `${request.toJSON().data} response: ${
              Buffer.from(message.data).toJSON().data
            }`
          );

          delete requests[requestKey];
          clearTimeout(timeoutId);

          resolve(message.data);
          parser.off("data", onData);
        }
      }

      parser.on("data", onData);
    });

    requests[requestKey] = dataRequest;

    log(`Writing ${request.toJSON().data} to ${port}`);
    serial.write(request);
    connection.bytesWritten += Buffer.byteLength(request);
  }

  // make every DataView unique to each request, even though
  // they are accessing the same set of data
  return dataRequest.then((responseData) => new MspDataView(responseData));
};

/**
 * Private, used for testing
 */
export const raw = (port: string): SerialPort | undefined =>
  connectionsMap[port]?.serial;

export const reset = (): void => {
  Object.keys(connectionsMap).forEach((port) => {
    delete connectionsMap[port];
  });
};

export const ports = async (): Promise<string[]> => {
  if (!SerialPort.Binding) {
    await initialiseBindings();
  }
  return SerialPort.list().then((data) => data.map(({ path }) => path));
};

/**
 * Close the given port
 */
export const close = async (port: string): Promise<void> => {
  const connection = connectionsMap[port];
  if (!connection) {
    return;
  }
  const { serial } = connection;

  const closePromise = new Promise((resolve) => serial.on("close", resolve));
  serial.close();
  delete connectionsMap[port];

  await closePromise;
};

/**
 * Open a serial connection on the given
 * port, onClose will be emited once the
 * connection is closed
 */
export const open: OpenConnectionFunction = async (
  port: string,
  options?: ConnectionOptions | OnCloseCallback,
  onClose?: OnCloseCallback
) => {
  if (connectionsMap[port]) {
    throw new Error(`${port} is already open`);
  }
  log(`opening ${port}`);

  // Handle overloads
  let onCloseCallback = onClose;
  let connectionOptions = options;

  if (typeof connectionOptions === "function") {
    onCloseCallback = connectionOptions;
    connectionOptions = {};
  }

  if (!SerialPort.Binding) {
    await initialiseBindings();
  }

  const serial = new SerialPort(port, {
    baudRate: 115200,
    ...connectionOptions,
    autoOpen: false,
  });

  await new Promise((resolve, reject) => {
    serial.open((err) => {
      if (err) {
        log(`error opening ${port}`);
        reject(err);
      } else {
        resolve();
      }
    });
  });

  const parser = serial.pipe(new MspParser());
  parser.setMaxListeners(1000000);

  const connection = {
    serial,
    parser,
    requests: {},
    bytesRead: 0,
    bytesWritten: 0,
    packetErrors: 0,
    mspInfo: {
      apiVersion: "0",
      mspProtocolVersion: 0,
    },
  };

  connectionsMap[port] = connection;

  log(`reading API version from ${port}`);
  try {
    const response = await execute(port, { code: 1 });
    connection.mspInfo = {
      mspProtocolVersion: response.readU8(),
      apiVersion: `${response.readU8()}.${response.readU8()}.0`,
    };

    log(`read apiVersion=${connection.mspInfo.apiVersion}`);
  } catch (e) {
    log(`failed to read MSP info from ${port}, failed to open`);
    await close(port);
    throw new Error(`Could not read MSP version from ${port}`);
  }

  // valid connection, setup listeners
  serial.on("data", (data: Buffer) => {
    connection.bytesRead += Buffer.byteLength(data);
  });

  serial.on("error", () => {
    log(`${port} on error received`);
    close(port);
  });

  serial.on("close", () => {
    log(`${port} on close received`);
    close(port);
    onCloseCallback?.();
  });

  connection.bytesRead = 0;
  connection.bytesWritten = 0;
  connection.packetErrors = 0;
};

export const connections = (): string[] => Object.keys(connectionsMap);
export const isOpen = (port: string): boolean => !!connectionsMap[port];

export const bytesRead = (port: string): number =>
  connectionsMap[port]?.bytesRead ?? 0;

export const bytesWritten = (port: string): number =>
  connectionsMap[port]?.bytesWritten ?? 0;

export const packetErrors = (port: string): number =>
  connectionsMap[port]?.packetErrors ?? 0;

export const apiVersion = (port: string): string =>
  connectionsMap[port]?.mspInfo.apiVersion ?? "0.0.0";
