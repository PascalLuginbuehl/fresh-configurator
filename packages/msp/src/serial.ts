/**
 * MSP Serial connections. This allows us to make multiple
 * serial connections, and enables execution of commands
 * on the given connection.
 */
import { MspDataView } from "./utils";
import { MspMessage, MspParser } from "./parser";
import SerialPort = require("serialport");
import { encode_message_v2 } from "./encoders";

interface Connection {
  serial: SerialPort;
  parser: MspParser;
}
const connectionsMap: Record<string, Connection> = {};

export const ports = () =>
  SerialPort.list().then(data => data.map(({ path }) => path));

export const connect = async (
  port: string,
  onClose?: () => void
): Promise<boolean> => {
  if (connectionsMap[port]) {
    return true;
  }

  let serial: SerialPort | undefined;
  await new Promise((resolve, reject) => {
    serial = new SerialPort(port, err => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });

  if (serial) {
    const parser = serial.pipe(new MspParser());
    parser.setMaxListeners(1000000);

    connectionsMap[port] = {
      serial,
      parser
    };

    serial.on("close", () => {
      delete connectionsMap[port];
      onClose?.();
    });
  }

  return true;
};

export const connections = (): string[] => Object.keys(connectionsMap);
export const isConnected = (port: string) => !!connectionsMap[port];

interface MspCommand {
  code: number;
  data?: number[];
  timeout?: number;
}

/**
 * Execute the given MspCommand on the given port.
 *
 * Return the data received from the command
 */
export const execute = async (
  port: string,
  { code, data, timeout = 5000 }: MspCommand
): Promise<MspDataView> => {
  if (!connectionsMap[port]) {
    throw new Error(`${port} is not open`);
  }

  const { parser, serial } = connectionsMap[port];

  serial.write(Buffer.from(encode_message_v2(code, data)));

  return new Promise((resolve, reject) => {
    const onData = (message: MspMessage) => {
      if (message.code === code) {
        // Copy the data view
        resolve(new MspDataView(message.dataView.buffer()));
        parser.off("data", onData);
      }
    };

    // Throw an error if timeout is reached
    setTimeout(() => {
      parser.off("data", onData);
      reject(new Error("timed out during execution"));
    }, timeout);

    parser.on("data", onData);
  });
};

/**
 * Close the given port
 */
export const close = async (port: string): Promise<void> => {
  connectionsMap[port]?.serial.close();
};
