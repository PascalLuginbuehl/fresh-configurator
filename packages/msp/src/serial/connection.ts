/**
 * MSP Serial connections. This allows us to make multiple
 * serial connections, and enables execution of commands
 * on the given connection.
 *
 * The module handles making requests, and dealing with only
 * making one of the same request at a time.
 */

import SerialPort from "serialport";
import debug from "debug";
import { MspDataView } from "./utils";
import { MspMessage, MspParser } from "./parser";
import { encodeMessageV2, encodeMessageV1 } from "./encoders";

const log = debug("connection");

interface Connection {
  serial: SerialPort;
  parser: MspParser;
  requests: Record<string, Promise<ArrayBuffer> | undefined>;
}
const connectionsMap: Record<string, Connection> = {};

/**
 * Private, used for testing
 */
export const raw = (port: string): SerialPort | undefined =>
  connectionsMap[port]?.serial;

export const reset = (): void => {
  Object.keys(connectionsMap).forEach(port => {
    delete connectionsMap[port];
  });
};

export const ports = (): Promise<string[]> =>
  SerialPort.list().then(data => data.map(({ path }) => path));

/**
 * Close the given port
 */
export const close = async (port: string): Promise<void> => {
  if (!connectionsMap[port]) {
    return;
  }
  const { serial } = connectionsMap[port];

  const closePromise = new Promise(resolve => serial.on("close", resolve));
  serial.close();
  delete connectionsMap[port];

  await closePromise;
};

/**
 * Open a serial connection on the given
 * port, onClose will be emited once the
 * connection is closed
 */
export const open = async (
  port: string,
  onClose?: () => void
): Promise<void> => {
  if (connectionsMap[port]) {
    throw new Error(`${port} is already open`);
  }

  const serial = new SerialPort(port, { autoOpen: false });

  await new Promise((resolve, reject) => {
    serial.open(err => {
      if (err) {
        log(`error opening ${port}`);
        reject(err);
      } else {
        resolve();
      }
    });
  });

  const parser = serial.pipe(new MspParser());
  parser.setMaxListeners(1000000);

  connectionsMap[port] = {
    serial,
    parser,
    requests: {}
  };

  serial.on("error", () => {
    log(`${port} on error received`);
    close(port);
  });

  serial.on("close", () => {
    log(`${port} on close received`);
    close(port);
    onClose?.();
  });
};

export const connections = (): string[] => Object.keys(connectionsMap);
export const isOpen = (port: string): boolean => !!connectionsMap[port];

interface MspCommand {
  code: number;
  data?: Buffer;
  timeout?: number;
}

/**
 * Execute the given MspCommand on the given port.
 *
 * Return the data received from the command
 */
export const execute = async (
  port: string,
  { code, data, timeout = 5000 }: MspCommand
): Promise<MspDataView> => {
  if (!connectionsMap[port]) {
    throw new Error(`${port} is not open`);
  }
  const { parser, serial, requests } = connectionsMap[port];

  const request =
    code > 254 ? encodeMessageV2(code, data) : encodeMessageV1(code, data);
  const requestKey = request.toString("utf-8");
  const existingRequest = requests[requestKey];

  // Prevent FC lockup by checking if this request is currently being performed
  // and if not, making the request
  if (!existingRequest) {
    requests[requestKey] = new Promise((resolve, reject) => {
      // Throw an error if timeout is reached
      const timeoutId = setTimeout(() => {
        delete requests[requestKey];
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        parser.off("data", onData);
        reject(new Error("Timed out during execution"));
      }, timeout);

      function onData(message: MspMessage): void {
        if (message.code === code && !message.crcError) {
          log(`${request.toJSON().data} response: ${message.data}`);

          delete requests[requestKey];
          clearTimeout(timeoutId);

          resolve(message.data);
          parser.off("data", onData);
        }
      }

      parser.on("data", onData);
    });

    log(`Writing ${request.toJSON().data} to ${port}`);
    serial.write(request);
  }

  // make every DataView unique to each request, even though
  // they are accessing the same set of data
  return requests[requestKey]!.then(
    responseData => new MspDataView(responseData)
  );
};
