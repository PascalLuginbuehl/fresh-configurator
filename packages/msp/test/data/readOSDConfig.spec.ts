import { execute, apiVersion } from "../../src/serial/connection";
import MspDataView from "../../src/serial/dataview";
import readOsdData from "./readOSDConfig.json";
import { readOSDConfig } from "../../src";
import codes from "../../src/serial/codes";

jest.mock("../../src/serial/connection");
const mockExecute = (execute as unknown) as jest.MockedFunction<typeof execute>;
const mockApiVersion = (apiVersion as unknown) as jest.MockedFunction<
  typeof apiVersion
>;

const simulateResponse = (data: number[]): void => {
  mockExecute.mockResolvedValue(new MspDataView(new Uint8Array(data).buffer));
};

describe("readOSDConfig", () => {
  it("should read a 1.42.0 config correctly", async () => {
    simulateResponse(readOsdData);
    mockApiVersion.mockReturnValue("1.42.0");

    const config = await readOSDConfig("/dev/something");
    expect(config).toMatchSnapshot();
    expect(mockExecute).toHaveBeenCalledWith("/dev/something", {
      code: codes.MSP_OSD_CONFIG,
    });
  });
});
