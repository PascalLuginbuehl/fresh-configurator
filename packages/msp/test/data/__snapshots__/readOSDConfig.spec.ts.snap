// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readOSDConfig should read a 1.42.0 config correctly 1`] = `
Object {
  "alarms": Array [
    Object {
      "key": 0,
      "value": 25,
    },
    Object {
      "key": 1,
      "value": 2200,
    },
    Object {
      "key": 3,
      "value": 100,
    },
  ],
  "displayItems": Array [
    Object {
      "key": 0,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 1,
      "position": Array [
        22,
        11,
      ],
      "visibility": Array [
        true,
        false,
        false,
      ],
    },
    Object {
      "key": 2,
      "position": Array [
        13,
        6,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 3,
      "position": Array [
        14,
        2,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 4,
      "position": Array [
        14,
        6,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 18,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 19,
      "position": Array [
        1,
        11,
      ],
      "visibility": Array [
        true,
        false,
        false,
      ],
    },
    Object {
      "key": 20,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 14,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 6,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 8,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 12,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 13,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 21,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 22,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 15,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 23,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 24,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 25,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 26,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 27,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 28,
      "position": Array [
        9,
        9,
      ],
      "visibility": Array [
        true,
        true,
        true,
      ],
    },
    Object {
      "key": 30,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 31,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 32,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 33,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 34,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 35,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 36,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 11,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 37,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 38,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 39,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 40,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 41,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 42,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 43,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 44,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 45,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 46,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 47,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 48,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 49,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 50,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 51,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 52,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 53,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 54,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 55,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 56,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 57,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 58,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 59,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 60,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 61,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        false,
        false,
        false,
      ],
    },
    Object {
      "key": 62,
      "position": Array [
        10,
        7,
      ],
      "visibility": Array [
        true,
        false,
        false,
      ],
    },
  ],
  "flags": Object {
    "hasOSD": true,
    "haveMax7456Video": true,
    "haveOsdFeature": true,
    "isMax7456Detected": true,
    "isOsdSlave": false,
  },
  "osdProfiles": Object {
    "count": 3,
    "selected": 1,
  },
  "parameters": Object {
    "cameraFrameHeight": 11,
    "cameraFrameWidth": 24,
    "overlayRadioMode": 2,
  },
  "staticItems": Array [
    Object {
      "enabled": false,
      "key": 12,
    },
    Object {
      "enabled": false,
      "key": 8,
    },
    Object {
      "enabled": true,
      "key": 9,
    },
    Object {
      "enabled": true,
      "key": 0,
    },
    Object {
      "enabled": false,
      "key": 10,
    },
    Object {
      "enabled": true,
      "key": 1,
    },
    Object {
      "enabled": false,
      "key": 7,
    },
    Object {
      "enabled": false,
      "key": 13,
    },
    Object {
      "enabled": true,
      "key": 2,
    },
    Object {
      "enabled": true,
      "key": 3,
    },
    Object {
      "enabled": true,
      "key": 4,
    },
    Object {
      "enabled": false,
      "key": 5,
    },
    Object {
      "enabled": true,
      "key": 6,
    },
    Object {
      "enabled": true,
      "key": 11,
    },
    Object {
      "enabled": false,
      "key": 14,
    },
    Object {
      "enabled": false,
      "key": 15,
    },
    Object {
      "enabled": false,
      "key": 16,
    },
    Object {
      "enabled": false,
      "key": 17,
    },
    Object {
      "enabled": false,
      "key": 18,
    },
    Object {
      "enabled": false,
      "key": 19,
    },
    Object {
      "enabled": false,
      "key": 20,
    },
    Object {
      "enabled": false,
      "key": 21,
    },
    Object {
      "enabled": false,
      "key": 22,
    },
    Object {
      "enabled": false,
      "key": 23,
    },
  ],
  "timerSources": Array [
    0,
    1,
    2,
    3,
  ],
  "timers": Array [
    Object {
      "key": 0,
      "precision": 0,
      "src": 0,
      "time": 10,
    },
    Object {
      "key": 1,
      "precision": 0,
      "src": 1,
      "time": 10,
    },
  ],
  "unitMode": 1,
  "videoSystem": 2,
  "warnings": Array [
    Object {
      "enabled": true,
      "key": 0,
    },
    Object {
      "enabled": true,
      "key": 1,
    },
    Object {
      "enabled": true,
      "key": 2,
    },
    Object {
      "enabled": true,
      "key": 3,
    },
    Object {
      "enabled": true,
      "key": 4,
    },
    Object {
      "enabled": true,
      "key": 5,
    },
    Object {
      "enabled": true,
      "key": 6,
    },
    Object {
      "enabled": true,
      "key": 7,
    },
    Object {
      "enabled": true,
      "key": 8,
    },
    Object {
      "enabled": true,
      "key": 9,
    },
    Object {
      "enabled": true,
      "key": 10,
    },
    Object {
      "enabled": true,
      "key": 11,
    },
    Object {
      "enabled": true,
      "key": 12,
    },
    Object {
      "enabled": false,
      "key": 13,
    },
    Object {
      "enabled": false,
      "key": 14,
    },
    Object {
      "enabled": false,
      "key": 15,
    },
  ],
}
`;
